app/models/date_validation.rb

```diff
@@ -26,6 +26,10 @@ def pre_validate_date_attribute(attribute, date)
     @invalid_date_attributes = Set.new if @invalid_date_attributes.nil?
     if date.is_a?(Hash)
       begin
+        # Rails will cast the year part of the date to 0 if the year input parameter is a non-numeric string
+        # This only seems to happen to the year part, other parts remain as strings
```

Comment:
Just to explain why I think this is needed. I don't think it's specifically doing anything special for 'year'.

My understanding is that Rails calls `.to_i` on every value in the hash, so they're all cast to integers. None of them will remain as strings.

As you said, non-numeric strings will become `0`. In other words:

```ruby
"5".to_i == 5 # => true
"0".to_i == 0 # => true
"abc".to_i == 0 # => true
```

However, the `Date.new` constructor only accepts `0` for the year parameter:

- `0` is a valid year (as is any other integer, positive **or** negative)
- `0` **is not** a valid month (must be >= 1)
- `0` **is not** a valid day (must be >= 1)

I think this is why you only saw this causing problems for the year. Because non-numeric strings for month and day would have triggered a `TypeError` on the Date constructor already – whereas `0` is a valid value for the year.

Reply:
Oh, that makes sense. I was trying to read through the attribute assignment code in Rails but it was a bit tricky to follow. Thanks for clarifying